README.md
# bedrock-package-manager

## Usage
A package should register itself using a Bedrock event handler.
```js
bedrock.events.on('bedrock.init', () => {
  brPackageManager.register({
    alias: 'mockPackage',
    meta: {
      optional: 'meta values',
    },
    packageName: '@bedrock/package-manager-mock-package',
    type: 'bedrock-mock-plugin',
  });
});
```

Use a registered package in an API.
```js
exports.getReports = async ({query = {}, storageApi}) => {
  // storageApi === 'mockPackage' would match the registration shown above
  const pkg = brPackageManager.get({
    alias: storageApi, type: 'bedrock-mock-plugin'
  });
  const storage = await import(pkg.packageName);
  return storage.find(query);
};
```

## API Reference
## Modules

<dl>
<dt><a href="#module_bedrock-package-manager">bedrock-package-manager</a></dt>
<dd></dd>
</dl>

## Typedefs

<dl>
<dt><a href="#PackageInfo">PackageInfo</a> : <code>Object</code></dt>
<dd><p>Package information.</p>
</dd>
</dl>

<a name="module_bedrock-package-manager"></a>

## bedrock-package-manager

* [bedrock-package-manager](#module_bedrock-package-manager)
    * [.get(options)](#module_bedrock-package-manager.get) ⇒ [<code>PackageInfo</code>](#PackageInfo)
    * [.findType(options)](#module_bedrock-package-manager.findType) ⇒ [<code>Array.&lt;PackageInfo&gt;</code>](#PackageInfo)
    * [.register(options)](#module_bedrock-package-manager.register) ⇒ <code>undefined</code>

<a name="module_bedrock-package-manager.get"></a>

### bedrock-package-manager.get(options) ⇒ [<code>PackageInfo</code>](#PackageInfo)
Get package information. One of `alias` or `packageName` is required.

**Kind**: static method of [<code>bedrock-package-manager</code>](#module_bedrock-package-manager)
**Returns**: [<code>PackageInfo</code>](#PackageInfo) - The package information.
**Throws**:

- <code>BedrockError</code> Will throw a `NotFoundError` if the package is not
  found.


| Param | Type | Description |
| --- | --- | --- |
| options | <code>Object</code> | The options to use. |
| options.type | <code>string</code> | The package type. |
| [options.alias] | <code>string</code> | The package alias. |
| [options.packageName] | <code>string</code> | The package name. |

<a name="module_bedrock-package-manager.findType"></a>

### bedrock-package-manager.findType(options) ⇒ [<code>Array.&lt;PackageInfo&gt;</code>](#PackageInfo)
Find packages by type.

**Kind**: static method of [<code>bedrock-package-manager</code>](#module_bedrock-package-manager)
**Returns**: [<code>Array.&lt;PackageInfo&gt;</code>](#PackageInfo) - The package information.

| Param | Type | Description |
| --- | --- | --- |
| options | <code>Object</code> | The options to use. |
| options.type | <code>string</code> | The package type to find. |

<a name="module_bedrock-package-manager.register"></a>

### bedrock-package-manager.register(options) ⇒ <code>undefined</code>
Register a package.

**Kind**: static method of [<code>bedrock-package-manager</code>](#module_bedrock-package-manager)
**Throws**:

- <code>BedrockError</code> Will throw a `DuplicateError` if the package is
  already registered.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| options | <code>Object</code> |  | The options to use. |
| options.packageName | <code>string</code> |  | The NPM package name. |
| options.type | <code>string</code> |  | The package type. |
| [options.meta] | <code>Object</code> | <code>{}</code> | Domain specific meta data. |
| [options.alias] | <code>string</code> |  | The package alias. A short name for the   package (e.g. 'mongodb', 'redis'). |

<a name="PackageInfo"></a>

## PackageInfo : <code>Object</code>
Package information.

**Kind**: global typedef
**Properties**

| Name | Type | Description |
| --- | --- | --- |
| alias | <code>string</code> | The package alias. A short name for the   package (e.g. 'mongodb', 'redis'). |
| packageName | <code>string</code> | The NPM package name. |
| type | <code>string</code> | The package type. |
| meta | <code>Object</code> | Domain specific meta data. |



CHANGELOG.md
# bedrock-package-manager

## 3.0.0 - 2022-04-28

### Changed
- **BREAKING**: Update peer deps:
  - `@bedrock/core@6`.

## 2.0.0 - 2022-04-02

### Changed
- **BREAKING**: Rename package to `@bedrock/package-manager`.
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Remove default export.
- **BREAKING**: Require node 14.x.

## 1.2.0 - 2022-03-28

### Added
- Setup coverage and github workflows.

### Changed
- Update peer deps:
  - `bedrock@4.5`.
- Update internals to use esm style and use `esm.js` to
  transpile to CommonJS.

## 1.1.0 - 2021-07-23

### Changed
- Update peer dependencies; use bedrock@4.

## 1.0.1 - 2019-11-12

### Changed
- Update dependencies.

## 1.0.0 - 2019-03-21

- See git history for changes previous to this release.


test/test.js
/*!
 * Copyright (c) 2019-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as bedrock from '@bedrock/core';
import '@bedrock/package-manager';
import '@bedrock/package-manager-mock-package';
import '@bedrock/test';

bedrock.start();


test/mocha/01-api.js
/*!
 * Copyright (c) 2019-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as brPackageManager from '@bedrock/package-manager';

describe('bedrock-package-manager', () => {
  describe('get API', () => {
    it('returns package information given packageName', async () => {
      // pluginName would typically be derived from the bedrock config or
      // from an HTTP route param
      const packageName = '@bedrock/package-manager-mock-package';
      const type = 'bedrock-mock-plugin';
      const result = brPackageManager.get({packageName, type});
      const mockResult = {
        alias: 'mockPackage',
        meta: {optional: 'meta values'},
        packageName,
        type
      };
      should.exist(result);
      result.should.eql(mockResult);
    });
    it('returns package information given alias', async () => {
      // pluginName would typically be derived from the bedrock config or
      // from an HTTP route param
      const alias = 'mockPackage';
      const packageName = '@bedrock/package-manager-mock-package';
      const type = 'bedrock-mock-plugin';
      const result = brPackageManager.get({alias, type});
      const mockResult = {
        alias: 'mockPackage',
        meta: {optional: 'meta values'},
        packageName,
        type
      };
      should.exist(result);
      result.should.eql(mockResult);
    });
    it('packageInfo is frozen and does not allow mutation', async () => {
      const alias = 'mockPackage';
      const type = 'bedrock-mock-plugin';
      const result = brPackageManager.get({alias, type});
      let error;
      try {
        result.meta = {newMeta: 'information'};
      } catch(e) {
        error = e;
      }
      should.exist(error);
      error.should.be.instanceof(TypeError);
      // meta should be unchanged from the original
      result.meta.should.eql({optional: 'meta values'});
    });
    it('throws NotFoundError on unknown package', async () => {
      let error;
      let result;
      const packageName = 'bedrock-unknown-plugin';
      try {
        result = brPackageManager.get({
          packageName,
          type: 'bedrock-mock-plugin'
        });
      } catch(e) {
        error = e;
      }
      should.not.exist(result);
      should.exist(error);
      error.name.should.equal('NotFoundError');
    });
  }); // end get API

  describe('findType API', () => {
    it('should return mock package', async () => {
      const type = 'bedrock-mock-plugin';
      const result = brPackageManager.findType({type});
      result.should.be.an('array');
      result.should.have.length(1);
      result[0].packageName.should.equal(
        '@bedrock/package-manager-mock-package');
    });
    it('should empty array on unknown type', async () => {
      const type = 'bedrock-unknown-type';
      const result = brPackageManager.findType({type});
      result.should.be.an('array');
      result.should.have.length(0);
    });
  }); // end getType API

  describe('register API', () => {
    it('throws DuplicateError on duplicate type and packageName', () => {
      let result;
      let error;
      try {
        // same type and packageName as mock-package
        result = brPackageManager.register({
          packageName: '@bedrock/package-manager-mock-package',
          type: 'bedrock-mock-plugin',
        });
      } catch(e) {
        error = e;
      }
      should.not.exist(result);
      should.exist(error);
      error.name.should.equal('DuplicateError');
    });
    it('throws DuplicateError on duplicate type and alias', () => {
      let result;
      let error;
      try {
        // alias and type is the same as mock-package, but packageName is unique
        result = brPackageManager.register({
          alias: 'mockPackage',
          packageName: 'bedrock-unique-package',
          type: 'bedrock-mock-plugin',
        });
      } catch(e) {
        error = e;
      }
      should.not.exist(result);
      should.exist(error);
      error.name.should.equal('DuplicateError');
    });
    it('successfully registers same packageName with a unique type', () => {
      let result;
      let error;
      try {
        // same packageName as mock-package, but different type
        result = brPackageManager.register({
          packageName: '@bedrock/package-manager-mock-package',
          type: 'bedrock-mock-plugin-unique1',
        });
      } catch(e) {
        error = e;
      }
      assertNoError(error);
      // no return value on success
      should.not.exist(result);
    });
    it('successfully registers same alias with a unique type', () => {
      let result;
      let error;
      try {
        // same packageName and alias as mock-package, but different type
        result = brPackageManager.register({
          alias: 'mockPackage',
          packageName: '@bedrock/package-manager-mock-package',
          type: 'bedrock-mock-plugin-unique2',
        });
      } catch(e) {
        error = e;
      }
      assertNoError(error);
      // no return value on success
      should.not.exist(result);
    });
  }); // end registerAPI
});


test/mock-package/index.js
/*!
 * Copyright (c) 2019-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as bedrock from '@bedrock/core';
import * as brPackageManager from '@bedrock/package-manager';

bedrock.events.on('bedrock.init', () => {
  brPackageManager.register({
    alias: 'mockPackage',
    meta: {
      optional: 'meta values',
    },
    packageName: '@bedrock/package-manager-mock-package',
    type: 'bedrock-mock-plugin',
  });
});

export const mockValue = 42;


lib/main.js
/*!
 * Copyright (c) 2019-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as bedrock from 'bedrock';
import assert from 'assert-plus';
import deepFreeze from 'deep-freeze';
const {util: {BedrockError}} = bedrock;

const packages = [];

/**
 * @module bedrock-package-manager
 */

/**
 * Get package information. One of `alias` or `packageName` is required.
 *
 * @param {object} options - The options to use.
 * @param {string} options.type - The package type.
 * @param {string} [options.alias] - The package alias.
 * @param {string} [options.packageName] - The package name.
 *
 * @returns {PackageInfo} The package information.
 *
 * @throws {BedrockError} Will throw a `NotFoundError` if the package is not
 *   found.
 */
export function get({alias, packageName, type}) {
  assert.optionalString(alias, 'options.alias');
  assert.optionalString(packageName, 'options.packageName');
  assert.string(type, 'options.type');

  let pkg;
  if(alias) {
    pkg = packages.find(p => (p.type === type && p.alias === alias));
  }
  if(packageName) {
    pkg = packages.find(p => (p.type === type &&
      p.packageName === packageName));
  }

  if(!pkg) {
    throw new BedrockError('Package not found.', 'NotFoundError', {
      alias,
      httpStatusCode: 404,
      packageName,
      public: true,
      type,
    });
  }
  return pkg;
}

/**
 * Find packages by type.
 *
 * @param {object} options - The options to use.
 * @param {string} options.type - The package type to find.
 *
 * @returns {PackageInfo[]} The package information.
 */
export function findType({type}) {
  assert.string(type, 'options.type');
  return packages.filter(p => p.type === type);
}

/**
 * Register a package.
 *
 * @param {object} options - The options to use.
 * @param {string} options.packageName - The NPM package name.
 * @param {string} options.type - The package type.
 * @param {object} [options.meta={}] - Domain specific meta data.
 * @param {string} [options.alias] - The package alias. A short name for the
 *   package (e.g. 'mongodb', 'redis').
 *
 * @returns {undefined}
 *
 * @throws {BedrockError} Will throw a `DuplicateError` if the package is
 *   already registered.
 */
export function register({alias, meta = {}, packageName, type}) {
  assert.optionalString(alias, 'options.alias');
  assert.string(packageName, 'options.name');
  assert.string(type, 'options.type');
  assert.object(meta, 'options.meta');

  if(packages.find(p => (p.packageName === packageName && p.type === type) ||
    (alias && p.alias === alias && p.type === type))) {
    throw new BedrockError('Duplicate package.', 'DuplicateError', {
      httpStatusCode: 409,
      packageName,
      public: true,
      type,
    });
  }

  const packageInfo = {alias, meta, packageName, type};
  // freeze the package information so that it cannot be altered after a call
  // to the `get` or `findType` APIs
  deepFreeze(packageInfo);
  packages.push(packageInfo);
}

/**
 * Package information.
 *
 * @typedef {object} PackageInfo
 * @property {string} alias - The package alias. A short name for the
 *   package (e.g. 'mongodb', 'redis').
 * @property {string} packageName - The NPM package name.
 * @property {string} type - The package type.
 * @property {object} meta - Domain specific meta data.
 */


lib/index.js
/*!
 * Copyright (c) 2019-2022 Digital Bazaar, Inc. All rights reserved.
 */
import * as bedrock from '@bedrock/core';
import assert from 'assert-plus';
import deepFreeze from 'deep-freeze';
const {util: {BedrockError}} = bedrock;

const packages = [];

/**
 * @module bedrock-package-manager
 */

/**
 * Get package information. One of `alias` or `packageName` is required.
 *
 * @param {object} options - The options to use.
 * @param {string} options.type - The package type.
 * @param {string} [options.alias] - The package alias.
 * @param {string} [options.packageName] - The package name.
 *
 * @returns {PackageInfo} The package information.
 *
 * @throws {BedrockError} Will throw a `NotFoundError` if the package is not
 *   found.
 */
export function get({alias, packageName, type}) {
  assert.optionalString(alias, 'options.alias');
  assert.optionalString(packageName, 'options.packageName');
  assert.string(type, 'options.type');

  let pkg;
  if(alias) {
    pkg = packages.find(p => (p.type === type && p.alias === alias));
  }
  if(packageName) {
    pkg = packages.find(p => (p.type === type &&
      p.packageName === packageName));
  }

  if(!pkg) {
    throw new BedrockError('Package not found.', 'NotFoundError', {
      alias,
      httpStatusCode: 404,
      packageName,
      public: true,
      type,
    });
  }
  return pkg;
}

/**
 * Find packages by type.
 *
 * @param {object} options - The options to use.
 * @param {string} options.type - The package type to find.
 *
 * @returns {PackageInfo[]} The package information.
 */
export function findType({type}) {
  assert.string(type, 'options.type');
  return packages.filter(p => p.type === type);
}

/**
 * Register a package.
 *
 * @param {object} options - The options to use.
 * @param {string} options.packageName - The NPM package name.
 * @param {string} options.type - The package type.
 * @param {object} [options.meta={}] - Domain specific meta data.
 * @param {string} [options.alias] - The package alias. A short name for the
 *   package (e.g. 'mongodb', 'redis').
 *
 * @returns {undefined}
 *
 * @throws {BedrockError} Will throw a `DuplicateError` if the package is
 *   already registered.
 */
export function register({alias, meta = {}, packageName, type}) {
  assert.optionalString(alias, 'options.alias');
  assert.string(packageName, 'options.name');
  assert.string(type, 'options.type');
  assert.object(meta, 'options.meta');

  if(packages.find(p => (p.packageName === packageName && p.type === type) ||
    (alias && p.alias === alias && p.type === type))) {
    throw new BedrockError('Duplicate package.', 'DuplicateError', {
      httpStatusCode: 409,
      packageName,
      public: true,
      type,
    });
  }

  const packageInfo = {alias, meta, packageName, type};
  // freeze the package information so that it cannot be altered after a call
  // to the `get` or `findType` APIs
  deepFreeze(packageInfo);
  packages.push(packageInfo);
}

/**
 * Package information.
 *
 * @typedef {object} PackageInfo
 * @property {string} alias - The package alias. A short name for the
 *   package (e.g. 'mongodb', 'redis').
 * @property {string} packageName - The NPM package name.
 * @property {string} type - The package type.
 * @property {object} meta - Domain specific meta data.
 */
